python/revng/ptml/wrappers.py

```diff
@@ -0,0 +1,129 @@
+YAMLLoader = getattr(yaml, "CSafeLoader", yaml.SafeLoader)
+
+Buffer = Union[str, bytes]
+FileLike = Union[IO[bytes], IO[str]]
+
+# When calling InputWrapper.keys the return is either
+# * A tuple with a lone `None`, indicating a single PTML document
+# * A list of string keys, for a multi-document container
+WrapperKeys = Union[Tuple[None], List[str]]
+
+
+class InputWrapper(Protocol):
+    """Wrapper class to seamlessly handle single and multi-document PTML
+    containers"""
+
+    def keys(self) -> WrapperKeys:
+        """Get the keys of the wrapper"""
+        ...
+
+    def get(self, key: Optional[str]) -> FileLike:
+        """Return a file-like object for the given key. Wrappers can throw an
+        exception if the key is not in `keys`."""
+        ...
+
+
+class TarWrapper(InputWrapper):
+    """Wrapper for PTML files contained in a .tar(.gz) file"""
+
+    def __init__(self, input_: BytesIO):
+        self.handle = tarfile.open(fileobj=input_)
+
+    def keys(self) -> List[str]:
+        return self.handle.getnames()
+
+    def get(self, key: Optional[str]) -> FileLike:
+        if key is None:
+            raise ValueError("Must specify a key")
+        result = self.handle.extractfile(key)
+        assert result is not None
+        return result
+
+
+class YamlWrapper(InputWrapper):
+    """Wrapper for a YAML dict with string keys and values containing a PTML
+    document"""
+
+    def __init__(self, input_: Buffer):
+        self.data: Dict[str, str] = yaml.load(input_, Loader=YAMLLoader)
+
+    def keys(self) -> List[str]:
+        return list(self.data.keys())
+
+    def get(self, key: Optional[str]) -> IO[str]:
+        if key is None:
+            raise ValueError("Must specify a key")
+        return StringIO(self.data[key])
+
+
+class SingleWrapper(InputWrapper):
+    """Wrapper for a single PTML document"""
+
+    def __init__(self, input_: Buffer):
+        self.data = input_
+
+    def keys(self) -> Tuple[None]:
+        return (None,)
+
+    def get(self, key: Optional[str]) -> FileLike:
+        if key is not None:
+            raise ValueError("Key must be None")
+        if isinstance(self.data, str):
+            return StringIO(self.data)
+        else:
+            return BytesIO(self.data)
+
+
+# This type encompasses the allowed inputs that are accepted by the functions
+# exposed in `revng.ptml`
+ParseInput = Union[str, bytes, IO[str], IO[bytes]]
+
+
+def _is_plain_ptml(input_: Union[str, bytes, mmap.mmap]) -> bool:
+    if isinstance(input_, str):
+        return re.match(r"\s*<", input_) is not None
+    else:
+        return re.match(rb"\s*<", input_) is not None
+
+
+def _wrap_inner(input_: Union[str, bytes, mmap.mmap]):
+    if len(input_) == 0:
+        raise ValueError("Input is empty!")
+
+    if isinstance(input_, (bytes, mmap.mmap)):
+        with suppress(tarfile.ReadError):
+            return TarWrapper(BytesIO(cast(bytes, input_)))
```

Comment:
Are we sure the best we can do here is trial and error?
Shouldn't we have the MIME type and avoid guessing?

Reply:
You're thinking on the pipeline side, there there's the mime.
In this context there's no mime, e.g.
```
with open("/some/random/file") as f:
  with wrap(f) as wrapped:
    _parse(f)
```